WIP: prototyping Graph/computing network.

rainwoodman opened the pull request.

rainwoodman committed:
stub for computing graph.

This problem is hard, because the graph shall be dynamically built.

rainwoodman committed:
Some doodling.

rainwoodman committed:
More doodling of the graphs.

I shall probably write a DummyExecutor for testing.

rainwoodman committed:
add a test.

rainwoodman committed:
A working example of computing a graph.

rainwoodman commented:
SubGraph : Graph as a node.

Named arguments?

arteymix commented:
I rebased this branch upon `master` so that you can have access to the function introspetion API.

arteymix commented:
I'll probably update this PR if you don't mind to use the new function introspection API.

I'll be in vacation for the coming week though, so don't expect anything before the 22 of december.

By the way, I think we should use a breadth-first evaluator with an async queue and some threads evaluating the graph in parallel. 

The idea is to push the variables to evaluate in the queue. Then, each thread pop the head and attempt to realize the node, if so, nice, you evaluate and store the result in the cache, otherwise push all the node dependencies in the queue then the node itself and iterate.

We also need a reuse strategy for items from the cache that become useless. This is why I think the slice allocator could be interesting there. We could use built-in refcounting to clean old variables and reallocate them using `Slice.alloc`.

I'll also write a chi-squared example if it works well. Do you have some sample data for that test case?

rainwoodman commented:
Sorry I've been trapped at work most of the week. 

For test data, I'd just fill the field with constant for a starter.

We probably still need a non-threaded walker.

arteymix commented:
Sure, in the end we'll probably have various flavours of executors anyway, so let's start with the simplest.

arteymix commented on src/graph.vala:
```diff
@@ -0,0 +1,307 @@
+    public Vast.Network.Node?
+    find_node(Vast.Network.Variable V)
+    {
+        /** FIXME: exception or null? **/
```
Always prefer returning `null` in these situations (e.g. finding an element in a data structure). Not having a node in the graph does not constitute an error condition and it's perfectly possible that no node generates the given variable in this case.

arteymix commented on src/graph.vala:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+public abstract class Vast.Network.Operation : Object
```
In the new model, the operation could simply be a `Vast.Function`.

arteymix commented on src/graph.vala:
```diff
@@ -0,0 +1,307 @@
+
+public class Vast.Network.Variable : Object
+{
+    public string name {get; construct set;}
```
Rather than having named variables with a counter we could use a UUID here.

arteymix committed:
Use UUIDs for variable names

arteymix committed:
graph: Improve APIs for flow graphs

Rework 'Operation' abstract class to perform operation on arrays. This
will be used as a base class for GI-based and OpenCL-based operations.
Move 'find_partial_derivative' there and remove the 'Gradient' class.

Rename 'Function' for 'GIOperation' to denote an introspected operation
via GIR.

Replace dummy variables with in/out direction flags.

Move 'Node' and 'Variable' into 'Vast.Graph' internal classes.

Make 'GraphExecutor' an abstract class and provide an iterative
implementation  with 'SimpleGraphExecutor'.

Provide asynchronous APIs to compute graph and invoke operations. This
is meant to be used along with gocl to schedule computation using the
'GLib.MainLoop'.

arteymix commented:
There's a really nice project called GOCL https://github.com/elima/gocl that provide a GObject wrapper around OpenCL.

It would be interesting to have a graph evaluator that would compile down the graph into a OpenCL kernel and evaluate it on a GPU or any compatible device.

arteymix commented:
This https://github.com/elima/gocl/pull/6 should allow us to map dense arrays directly into OpenCL devices.

If we follow a memory mapping approach, we can have a single implementation with different mapping strategy. I wonder if it's possible to have a custom mmap protocol to do stuff like implicit byteswapping.

arteymix committed:
Add 'SimpleOperation' for user-defined operations using callbacks

arteymix commented:
I need to fix the current evaluation algorithm because it loops on `inout` variables. I'm pretty sure we'll need directions on node as well.
